### ISSUE TITLE
Zed applies parent `.gitignore` files into open directory

### ISSUE THREAD

**Issue by rmachado-studocu** [zed-industries/zed#6998]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Zed applies parent `.gitignore` even if only the nested folder is open.

Steps to repro:

```bash
mkdir temp && cd temp
# creates a default nextjs app
yarn dlx create-next-app@latest my-app --ts --eslint --no-tailwind --no-src-dir --use-yarn --no-app --import-alias '@/*'
echo "my-app/**/*" >> .gitignore
zed my-app
```

Then `cmd+P` and try to search for `index.tsx` ... Cannot find it? Easy:

```bash
# This assumes you are in the `temp` directory and removes its `.gitignore`
rm .gitignore
```

Then `cmd+P` and try to search for `index.tsx` ... Now you can!

> [!IMPORTANT]
> Notice that the command to open the IDE is `zed my-app` which means that I have _no business_ with the `temp` folder... the `.gitgnore` should not be applied given that for the IDE the "project" only has the `my-app` folder and sub-folders ...

### Environment

Zed: v0.119.20 (Zed)
OS: macOS 13.3.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

https://github.com/zed-industries/zed/assets/89906313/e33a3e4e-5895-4fa7-9a7a-a03e4b319016

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

**Comment by GregdTd** [comment]
+1

**Comment by fdionisi** [comment]
By reading the code, it seems this is intended behaviour [^1]. Nevertheless, git’s gitignore documentation states as follows:
> Patterns read from a .gitignore file in the same directory as the path, or in any parent directory (up to the top-level of the working tree), with patterns in the higher level files being overridden by those in lower level files down to the directory containing the file. These patterns match relative to the location of the .gitignore file. A project normally includes such .gitignore files in its repository, containing patterns for files generated as part of the project build. [^2]

What I understand is that we should only consider `.gitignore` files up to the _“top-level of the working tree”_, hence, in considering ancestor folders until we find the root containing `.git/`.

Troubleshooting with git, seems to be indeed the expected behaviours.

```sh
git init test-project
mkdir -p test-project/services/first
echo 'services/*' > test-project/.gitignore

touch test-project/services/first/test.1 test-project/services/first/test.2
echo 'test.2' > test-project/services/first/.gitignore

cd test-project/services/first

git status
# On branch main
# 
# No commits yet
# 
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
# 	../../.gitignore
# 
# nothing added to commit but untracked files present (use "git add" to track)

git init
git status
# On branch main
# 
# No commits yet
# 
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
# 	      .gitignore
# 	      test.1
# 
# nothing added to commit but untracked files present (use "git add" to track)
```

VS Code instead, applies .gitignore files up until the open working tree. Which means:
1. Opening `test-project` or `test-project/services/first` applies top level .gitignore, matching git behaviour:
	
	<img width="710" alt="Screenshot 2024-02-21 at 8 33 06 PM" src="https://github.com/zed-industries/zed/assets/8927326/2a8ff056-0c03-480b-a7e7-145af4828d25">
	
	<img width="710" alt="Screenshot 2024-02-21 at 8 33 42 PM" src="https://github.com/zed-industries/zed/assets/8927326/40e578f4-d977-4a22-8a4b-9295b0badd33">

3. Opening `test-project/services` ignores files only based on the child `first/.gitignore`, while inviting the user to open the parent git folder.  
	<img width="710" alt="Screenshot 2024-02-21 at 8 33 20 PM" src="https://github.com/zed-industries/zed/assets/8927326/4c43ba62-af5b-4506-b21f-65b0bf6d3a64">

If we want to match the behaviour, we should remove the loop altogether [^1] and check the presence of `.git` in the root or parent when opening a work tree. 

@JosephTLyons let me know if that's the the direction, and I'd be happy to give it a try.

[^1]: https://github.com/zed-industries/zed/blob/aa6926e57ae78afc421dbdaabcab5907936d3d1e/crates/project/src/worktree.rs#L3320-L3331
[^2]: https://git-scm.com/docs/gitignore

**Comment by rmachado-studocu** [comment]
Well, VS Code's behaviour is quite simple to be tested...
Most common projects have node_modules installed and `.gitignore`'d  and yet if you open the `node_modules` in the IDE (not its parent directory) it'll not apply the .gitignore of the parent (nor it should, as it should not be _aware_ that there's a .gitignore in the parent... it's not its responsibility).

**Comment by fdionisi** [comment]
Indeed, you are right! Mine are only  some thoughts after looking at Zed’s source code. 😊

**Comment by JosephTLyons** [comment]
For anyone following this thread for this bug, can you try the latest preview version and see if the issue persists?

https://zed.dev/releases/preview

**Comment by fdionisi** [comment]
TL;DR Looks like the core issue has been solved, but the UX is a bit wonky.

Considering [my preproduction](https://github.com/zed-industries/zed/issues/6998#issuecomment-1957786515), when I first open the parent repository everything looks in order with the folders in `services/` ignored.

<img width="873" alt="Screenshot 2024-03-04 at 8 41 21 AM" src="https://github.com/zed-industries/zed/assets/8927326/f7813627-a87b-43c6-a2a7-77396ccbfb8a">

But when I open the ignored `services/first` project, files gets highlighted applying `services/first/.gitignore`

<img width="873" alt="Screenshot 2024-03-04 at 8 41 25 AM" src="https://github.com/zed-industries/zed/assets/8927326/eb7ad92c-6dc5-4e82-9cd1-006f1dfcdfe8">

That's different from git's behaviour, where the root level git applies.

<img width="1015" alt="Screenshot 2024-03-04 at 8 55 13 AM" src="https://github.com/zed-industries/zed/assets/8927326/69bbe2c3-f08f-456c-9142-e9bddf36cd55">

Fun fact, when running `git status` in the root, ignored children files in the tree view looses their green status, just to re-gain it on click.

**Comment by rmachado-studocu** [comment]
Yup, seems to be solved in Zed Preview 👍 Thank y'all for working on this!

**Comment by JosephTLyons** [comment]
> Yup, seems to be solved in Zed Preview 👍 Thank y'all for working on this!

Great! Let's close this out - we can reopen if it exists still in some form.

**Comment by neunato** [comment]
I'm still experiencing this issue on Linux.

Project files show in file finder, but no search results nor sidebar annotations.
